Allow help:evaluate output in quiet mode (MPH-99)

zigarn opened the pull request:
http://jira.codehaus.org/browse/MPH-99

zigarn committed:
Allow help:evaluate output in quiet mode (MPH-99)

jonwolski commented:
:+1:

djleeds commented:
ditto :+1:

robertmassaioli commented:
Agreed, the use case here is running help:evaluate in a bash script. I would actually recommend adding an extra CLI argument called 'ignoreQuiet' and, if that is true, then we ignore the quiet flag.

For example:

```
projectVersion=$(mvn -q help:evaluate -Dexpression=project.version)
```

It would be really handy to be able to do this.

alicederyn commented:
:+1:

thomas-p-wilson commented:
:+1:

michael-o commented:
Here is my proposal I would merge: Resemble the logic of the `effective` goals. I.e., provide an `output` parameter. It makes the stuff consistent.

zigarn commented:
For consistency, I agree that it's a good idea to add the `output` parameter.
But for usability, I think it's a good idea to write directly to stdout, and for consistency add this behavior to other mojos.
The `output` solution wouldn't be enough in case of interactive mode for `help:evaluate`.

michael-o commented:
@zigarn Can you elaborate on the last statement. I do not understand why it won't be enough.

zigarn commented:
Interactive mode through a file is not interactive. (I know it's a corner case to use interactive mode with `mvn --quiet`)

michael-o commented on maven-help-plugin/src/main/java/org/apache/maven/plugins/help/EvaluateMojo.java:
```diff
@@ -211,7 +211,7 @@ public void execute()
             while ( true )
             {
-                getLog().info( "Enter the Maven expression i.e. ${project.groupId} or 0 to exit?:" );
+                System.out.println( "Enter the Maven expression i.e. ${project.groupId} or 0 to exit?:" );
```
That is supposed to be `print` and not `println`. Otherwise you have an ugly line break.

michael-o commented:
I will have a look again, I think we can do even better which suits everyone.

zigarn commented:
Maybe allow `-` as a value for `output` ?

michael-o commented:
No, this is not what I was thinking about. I will take a quiet look at it later this day and update you.

michael-o commented:
Let's summarize the possible modes:
1. Interactive, parameter
2. Interactive, prompt
3. Interactive, parameter, quiet
4. Interactive, prompt, quiet
5. Batch, parameter
6. Batch, prompt
7. Batch, parameter, quiet
8. Batch, prompt, quiet

Now analyze:
1. Implemented
2. Implemented
3. Implemented
4. Implemented
5. Implemented
6. Does not make sense
7. Behaves like 3.
8. Does not make sense

We are talking about 2, 3 and 4.
1. Could definitively better with a `System.out.print`
2. \+ 4. quiet is quiet, where is the usecase for that? I agree that the current output style is crap.

The patch itself is faulty. See inline comments.

If we change the behavior of this goal, it won't be consistent with the rest anymore.

Comments welcome.

michael-o commented on maven-help-plugin/src/main/java/org/apache/maven/plugins/help/EvaluateMojo.java:
```diff
@@ -464,7 +464,7 @@ else if ( obj instanceof Settings )
 
-        getLog().info( "\n" + response.toString() );
+        System.out.println( response.toString() );
     }
```
You completely ignore batch mode.

zigarn commented:
1 Interactive, parameter
Without patch:

```
$ mvn help:evaluate -Dexpression="project.groupId"
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building Apache Maven Help Plugin 2.2.1-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-help-plugin:2.2:evaluate (default-cli) @ maven-help-plugin ---
[INFO] No artifact parameter specified, using 'org.apache.maven.plugins:maven-help-plugin:maven-plugin:2.2.1-SNAPSHOT' as project.
[INFO] 
org.apache.maven.plugins
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 2.160 s
[INFO] Finished at: 2015-09-30T11:26:43+02:00
[INFO] Final Memory: 15M/92M
[INFO] ------------------------------------------------------------------------
```

With patch:

```
$ mvn org.apache.maven.plugins:maven-help-plugin:2.2.1-SNAPSHOT:evaluate -Dexpression="project.groupId"
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building Apache Maven Help Plugin 2.2.1-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-help-plugin:2.2.1-SNAPSHOT:evaluate (default-cli) @ maven-help-plugin ---
[INFO] No artifact parameter specified, using 'org.apache.maven.plugins:maven-help-plugin:maven-plugin:2.2.1-SNAPSHOT' as project.
org.apache.maven.plugins
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1.993 s
[INFO] Finished at: 2015-09-30T11:32:42+02:00
[INFO] Final Memory: 16M/91M
[INFO] ------------------------------------------------------------------------
```

==> Only difference is the empty `[INFO]` line without patch (workaround of the current version to display in a new line using logging)
With the `output` parameter, the answer line would disappear and be written in the target file.

2 Interactive, prompt
Without patch:

```
$ mvn help:evaluate
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building Apache Maven Help Plugin 2.2.1-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-help-plugin:2.2:evaluate (default-cli) @ maven-help-plugin ---
[INFO] No artifact parameter specified, using 'org.apache.maven.plugins:maven-help-plugin:maven-plugin:2.2.1-SNAPSHOT' as project.
[INFO] Enter the Maven expression i.e. ${project.groupId} or 0 to exit?:
${project.groupId}
[INFO] 
org.apache.maven.plugins
[INFO] Enter the Maven expression i.e. ${project.groupId} or 0 to exit?:
0
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 10.670 s
[INFO] Finished at: 2015-09-30T11:27:57+02:00
[INFO] Final Memory: 15M/92M
[INFO] ------------------------------------------------------------------------
```

With patch:

```
$ mvn org.apache.maven.plugins:maven-help-plugin:2.2.1-SNAPSHOT:evaluate
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building Apache Maven Help Plugin 2.2.1-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-help-plugin:2.2.1-SNAPSHOT:evaluate (default-cli) @ maven-help-plugin ---
[INFO] No artifact parameter specified, using 'org.apache.maven.plugins:maven-help-plugin:maven-plugin:2.2.1-SNAPSHOT' as project.
Enter the Maven expression i.e. ${project.groupId} or 0 to exit?:
${project.groupId}
org.apache.maven.plugins
Enter the Maven expression i.e. ${project.groupId} or 0 to exit?:
0
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 10.205 s
[INFO] Finished at: 2015-09-30T11:33:54+02:00
[INFO] Final Memory: 16M/91M
[INFO] ------------------------------------------------------------------------
```

==> Only differences are the `[INFO]` on the prompt line and empty `[INFO]` line before the answer without patch. No more ugly line break with the patch than without it, but I agree that it could be better to wait for the input on the same line than the prompt.
The `output` parameter would be a non-sense in this case (it was my "interactive" notion in previous comments).

3 Interactive, parameter, quiet
Without patch:

```
$ mvn -q help:evaluate -Dexpression="project.groupId"
$ 
```

With patch:

```
$ mvn -q org.apache.maven.plugins:maven-help-plugin:2.2.1-SNAPSHOT:evaluate -Dexpression="project.groupId"
org.apache.maven.plugins
$ 
```

==> No output at all without patch and just the answer with the patch, allowing @robertmassaioli 's use case
The `output` parameter would be a compromise but the use case would be more complicated:

```
projectVersion=$(file=$(mktemp); mvn -q help:evaluate -Doutput=$file -Dexpression=project.version; cat $file; rm $file)
```

4 Interactive, prompt, quiet
Without patch:

```
$ mvn -q help:evaluate
${project.groupId}
0
$ 
```

With patch:

```
$ mvn -q org.apache.maven.plugins:maven-help-plugin:2.2.1-SNAPSHOT:evaluate
Enter the Maven expression i.e. ${project.groupId} or 0 to exit?:
${project.groupId}
org.apache.maven.plugins
Enter the Maven expression i.e. ${project.groupId} or 0 to exit?:
0
$ 
```

==> No prompt and no output without the patch. I agree it's a far-fetched case.

5 Batch, parameter
Same as 1.

6 Batch, prompt
No sense.
==> Is always rejected `[ERROR] Maven is configured to NOT interact with the user for input. This Mojo requires that 'interactiveMode' in your settings file is flag to 'true'.`

7 Batch, parameter, quiet
Same as 3.

8 Batch, prompt, quiet
No sense.
==> Is always rejected `[ERROR] Maven is configured to NOT interact with the user for input. This Mojo requires that 'interactiveMode' in your settings file is flag to 'true'.`

To compare with other plugins behavior:
- maven-release-plugin:
  - in prompt mode: like in the patch: all output in normal mode, only prompts in quiet mode
  - in batch mode: use default values

`help:effective-x` could benefit the same improvements by displaying on stdout without all the mvn logs output.

michael-o commented:
@zigarn Thank you for this quality response. I'll read it and will answer shortly.

michael-o commented:
@zigarn This took longer that I have expected but anyway, here it is:
1. patch is fine
2. patch is fine
3. patch is fine
4. patch is fine
5. That should use the logger instead of sysout
6. That should use the logger instead of sysout

One problem I see is:

```
getLog().warn( "The Maven expression was invalid. Please use a valid expression." );
return;
```

in `handleResponse` when interactive, quiet is used.

I see some inconsistencies and you might see. I am in favor to resolve them first.

michael-o commented:
Anyone?

zigarn commented:
I don't know why in batch mode it should use the logger.
For 5, the result will be the same with expection when using `--log-file` option.
For 7 (I guess, as 6 has no sense), using logging will have the current behavior of no output.

I updated the PR to incorporate your feedback.
I implemented the use of logger in batch mode in branch [MPH-99-batch-log](https://github.com/zigarn/maven-plugins/compare/MPH-99...zigarn:MPH-99-batch-log).

jglick commented on maven-help-plugin/src/main/java/org/apache/maven/plugins/help/EvaluateMojo.java:
```diff
@@ -373,7 +373,7 @@ private void handleResponse( String expr )
         if ( obj != null && expr.equals( obj.toString() ) )
         {
-            getLog().warn( "The Maven expression was invalid. Please use a valid expression." );
+            System.out.println( "The Maven expression was invalid. Please use a valid expression." );
```
Probably this and the previous one should use `System.err`, to avoid sending junk text to a program reading stdout.

michael-o replied:
As far as I know, no Maven sends output to `stderr`.  I am not going to break this convention.

rantoniuk commented:
Bump?

bondolo commented:
Bump indeed.

michael-o commented:
This is not forgotten and still on my TODO list with very little time.

rantoniuk commented:
Is any of the PRs mentioned in the comments above good enough?
(this one or [this one](https://github.com/zigarn/maven-plugins/compare/MPH-99...zigarn:MPH-99-batch-log) ) ?

Tunaki commented:
Writing info messages to the console is quiet mode is still awkward, as this mode [is documented](http://maven.apache.org/ref/3.3.9/maven-embedder/cli.html) only to output error messages.

I think adding the `output` parameter is a good idea, with this workflow:
- `mvn help:evaluate -Dexpression="project.groupId"` => logs the result in the console, like before
- `mvn help:evaluate -Dexpression="project.groupId" -Doutput=result.txt` => writes the result in the `result.txt` file.
- `mvn help:evaluate` => starts an interactive prompt, like before
- `mvn help:evaluate -Doutput=result.txt` => starts an interactive prompt, like before. `output` is ignored as it doesn't make sense to write to a file in an interactive prompt.

With this, the plugin behaves the same way as before, and respects the `quiet` documentation, but the wanted use-case of having just the result of the evaluation available is possible by specifying the expression and an output file. Keep in mind that you can also evaluate Maven projects or settings file, with `${project}` or `${settings}`; writing those to a file is cleaner. This use-case is only really useful for scripts calling `help:evaluate`, so having the result in a file that the script can read later on solves it. And outside of a script, one could still do `mvn -q help:evaluate -Dexpression="project.groupId" -Doutput=result.txt && type result.txt` (or `cat result.txt`) to just see the result.

Here's what it would look like with those changes:
1. `mvn help:evaluate`
   
   ```
   [INFO] Scanning for projects...
   [INFO]
   [INFO] ------------------------------------------------------------------------
   [INFO] Building Apache Maven Help Plugin 2.2.1-SNAPSHOT
   [INFO] ------------------------------------------------------------------------
   [INFO]
   [INFO] --- maven-help-plugin:2.2.1-SNAPSHOT:evaluate (default-cli) @ maven-help-plugin ---
   [INFO] No artifact parameter specified, using 'org.apache.maven.plugins:maven-help-plugin:maven-plugin:2.2.1-SNAPSHOT' as project.
   [INFO] Enter the Maven expression i.e. ${project.groupId} or 0 to exit?:
   ${project.groupId}
   [INFO]
   org.apache.maven.plugins
   [INFO] Enter the Maven expression i.e. ${project.groupId} or 0 to exit?:
   0
   [INFO] ------------------------------------------------------------------------
   [INFO] BUILD SUCCESS
   [INFO] ------------------------------------------------------------------------
   [INFO] Total time: 10.171 s
   [INFO] Finished at: 2016-10-16T18:36:59+02:00
   [INFO] Final Memory: 16M/164M
   [INFO] ------------------------------------------------------------------------
   ```
   
   Behaves the same as before.
2. `mvn help:evaluate -Doutput=result.txt`
   
   ```
   [INFO] Scanning for projects...
   [INFO]
   [INFO] ------------------------------------------------------------------------
   [INFO] Building Apache Maven Help Plugin 2.2.1-SNAPSHOT
   [INFO] ------------------------------------------------------------------------
   [INFO]
   [INFO] --- maven-help-plugin:2.2.1-SNAPSHOT:evaluate (default-cli) @ maven-help-plugin ---
   [INFO] No artifact parameter specified, using 'org.apache.maven.plugins:maven-help-plugin:maven-plugin:2.2.1-SNAPSHOT' as project.
   [WARNING] In interactive mode, the output will be written on the console, ignoring 'output'.
   [INFO] Enter the Maven expression i.e. ${project.groupId} or 0 to exit?:
   ${project.groupId}
   [INFO]
   org.apache.maven.plugins
   [INFO] Enter the Maven expression i.e. ${project.groupId} or 0 to exit?:
   0
   [INFO] ------------------------------------------------------------------------
   [INFO] BUILD SUCCESS
   [INFO] ------------------------------------------------------------------------
   [INFO] Total time: 10.945 s
   [INFO] Finished at: 2016-10-16T18:40:17+02:00
   [INFO] Final Memory: 16M/169M
   [INFO] ------------------------------------------------------------------------
   ```
   
   Behaves the same as before, and prints a warning to say that the `output` parameter was ignored, since in interactive mode, it doesn't make sense to print to a file. Possibly, we could also raise an error as the user probably wanted to set an expression as well.
3. `mvn help:evaluate -Dexpression=project.groupId`
   
   ```
   [INFO] Scanning for projects...
   [INFO]
   [INFO] ------------------------------------------------------------------------
   [INFO] Building Apache Maven Help Plugin 2.2.1-SNAPSHOT
   [INFO] ------------------------------------------------------------------------
   [INFO]
   [INFO] --- maven-help-plugin:2.2.1-SNAPSHOT:evaluate (default-cli) @ maven-help-plugin ---
   [INFO] No artifact parameter specified, using 'org.apache.maven.plugins:maven-help-plugin:maven-plugin:2.2.1-SNAPSHOT' as project.
   [INFO]
   org.apache.maven.plugins
   [INFO] ------------------------------------------------------------------------
   [INFO] BUILD SUCCESS
   [INFO] ------------------------------------------------------------------------
   [INFO] Total time: 1.312 s
   [INFO] Finished at: 2016-10-16T18:44:31+02:00
   [INFO] Final Memory: 16M/164M
   [INFO] ------------------------------------------------------------------------
   ```
   
   Behaves the same as before.
4. `mvn help:evaluate -Dexpression=project.groupId -Doutput=result.txt`
   
   ```
   [INFO] Scanning for projects...
   [INFO]
   [INFO] ------------------------------------------------------------------------
   [INFO] Building Apache Maven Help Plugin 2.2.1-SNAPSHOT
   [INFO] ------------------------------------------------------------------------
   [INFO]
   [INFO] --- maven-help-plugin:2.2.1-SNAPSHOT:evaluate (default-cli) @ maven-help-plugin ---
   [INFO] No artifact parameter specified, using 'org.apache.maven.plugins:maven-help-plugin:maven-plugin:2.2.1-SNAPSHOT' as project.
   [INFO] Result of evaluation written to: /path/to/result.txt
   [INFO] ------------------------------------------------------------------------
   [INFO] BUILD SUCCESS
   [INFO] ------------------------------------------------------------------------
   [INFO] Total time: 1.264 s
   [INFO] Finished at: 2016-10-16T18:45:54+02:00
   [INFO] Final Memory: 16M/164M
   [INFO] ------------------------------------------------------------------------
   ```
   
   Actual change: the result is not written in the console, but in the output file, as specified by the new `INFO` log line. The file would contain the single line: `org.apache.maven.plugins`.

In batch mode, point 1 and 2 above would raise an error, as expected. Point 3 and 4 would behave exactly the same way (specifying `output` writes the result to a file, otherwise, log to the console). This also integrates well with the `--log-file` parameter of Maven: point 1 and 2 above, in interactive mode, don't really make sense (nothing will be printed, as expected, just like before); point 3 and 4 would redirect just the logs to a file instead of the console, so it wouldn't interact at all with `output`.

Any feedback on that?

michael-o commented:
@Tunaki I am fully happy with you approach.

lalmeras commented:
Can you confirm that this feature is not currently availabel ? It seems there is no release of help plugin since 2013 ?

bondolo commented:
Some ability to redirect to stdout is essential for the script use-case. I've never wanted to send the result of a help:evaluate to a file and if I did can do so easily via existing mechanisms. Requiring that I send the result to a real file is just an annoyance.

bondolo commented:
New jira case for this same issue: https://issues.apache.org/jira/browse/MPH-144

khmarbaise commented:
A current solution would be to do it like this:
```
mvn help:evaluate -Dexpression=project.version | grep -v "\[INFO\]"
```
which is not ideal and would not work in some situations (if having a WARNING in your build)..which would result in the need to do it like this:
```
mvn help:evaluate -Dexpression=project.version | grep -v "\[INFO\]" | grep -v "\[WARNING\]"
```

Now we have implemented it in [MPH-144](https://issues.apache.org/jira/browse/MPH-144) which will reduce it to the following:
```
mvn help:evaluate -Dexpression=project.version -q -DforceStdout=true
```
without redirecting to a file etc..can also be used like this:
```
RESULT=$(mvn help:evaluate -Dexpression=project.version -q -DforceStdout=true)
```
It would be great if someone who likes/needs it can vote on the JIRA ticket (I see some watchers there but only a single VOTE on it)...So the chances are becoming better to integrate into the next release...

bondolo commented:
An example of what I am currently using in scripts:

    rversion=$(mvn -B help:evaluate -Dexpression=project.version | grep "^[^\\[]")

ie. I remove all lines which start with "["

j4b0l commented:
I would really like to see that fixed without workarounds, though...
There is a workaround - use Exec Maven Plugin, you can cover some of help:evaluate functionality running echo binary with parameter of your expression, e.g.
rversion=$(mvn -B -q \
    -Dexec.executable="echo" \
    -Dexec.args='${project.version}' \
    --non-recursive \
    org.codehaus.mojo:exec-maven-plugin:1.6.0:exec)

For some of expressions you'll have to run some other goals or lifecycle phases before. In most cases I need, it does the job.
